File: 1、Block原理/Block堆栈对变量捕获的影响/main.cpp

struct __block_impl {
    void *isa;
    int Flags;
    int Reserved;
    void *FuncPtr;
};

int gloable = 5;

#pragma mark - block1 __NSGlobalBlock__ -
struct __main_block_impl_0 {
    struct __block_impl impl;
    struct __main_block_desc_0* Desc;
    
    __main_block_impl_0(void *fp, struct __main_block_desc_0 *desc, int flags=0) {
        impl.isa = &_NSConcreteStackBlock;
        impl.Flags = flags;
        impl.FuncPtr = fp;
        Desc = desc;
    }
};
static void __main_block_func_0(struct __main_block_impl_0 *__cself) {
    
    NSLog((NSString *)&__NSConstantStringImpl__var_folders_c4_8r5g_7xx6szc_1fms0527v480000gn_T_main_46e938_mi_0);
}

static struct __main_block_desc_0 {
    size_t reserved;
    size_t Block_size;
} __main_block_desc_0_DATA = { 0, sizeof(struct __main_block_impl_0)};


#pragma mark - block2 __NSStackBlock__ -
struct __main_block_impl_1 {
    struct __block_impl impl;
    struct __main_block_desc_1* Desc;
    int age;
    __main_block_impl_1(void *fp, struct __main_block_desc_1 *desc, int _age, int flags=0) : age(_age) {
        impl.isa = &_NSConcreteStackBlock;
        impl.Flags = flags;
        impl.FuncPtr = fp;
        Desc = desc;
    }
};
static void __main_block_func_1(struct __main_block_impl_1 *__cself) {
    int age = __cself->age; // bound by copy
    
    NSLog((NSString *)&__NSConstantStringImpl__var_folders_c4_8r5g_7xx6szc_1fms0527v480000gn_T_main_46e938_mi_1);
    NSLog((NSString *)&__NSConstantStringImpl__var_folders_c4_8r5g_7xx6szc_1fms0527v480000gn_T_main_46e938_mi_2,age);
}

static struct __main_block_desc_1 {
    size_t reserved;
    size_t Block_size;
} __main_block_desc_1_DATA = { 0, sizeof(struct __main_block_impl_1)};


#pragma mark - block3 __NSMallocBlock__ -
struct __main_block_impl_2 {
    struct __block_impl impl;
    struct __main_block_desc_2* Desc;
    int age;
    __main_block_impl_2(void *fp, struct __main_block_desc_2 *desc, int _age, int flags=0) : age(_age) {
        impl.isa = &_NSConcreteStackBlock;
        impl.Flags = flags;
        impl.FuncPtr = fp;
        Desc = desc;
    }
};
static void __main_block_func_2(struct __main_block_impl_2 *__cself) {
    int age = __cself->age; // bound by copy
    
    NSLog((NSString *)&__NSConstantStringImpl__var_folders_c4_8r5g_7xx6szc_1fms0527v480000gn_T_main_46e938_mi_3);
    NSLog((NSString *)&__NSConstantStringImpl__var_folders_c4_8r5g_7xx6szc_1fms0527v480000gn_T_main_46e938_mi_4,age);
}

static struct __main_block_desc_2 {
    size_t reserved;
    size_t Block_size;
} __main_block_desc_2_DATA = { 0, sizeof(struct __main_block_impl_2)};


int main(int argc, const char * argv[]) {
    
    /* @autoreleasepool */ { __AtAutoreleasePool __autoreleasepool;
        
        void (*block1)(void) = ((void (*)())&__main_block_impl_0((void *)__main_block_func_0, &__main_block_desc_0_DATA));
        
        int age = 10;
        void (*block2)(void) = ((void (*)())&__main_block_impl_1((void *)__main_block_func_1, &__main_block_desc_1_DATA, age));
        
        void (*block3)(void) = (void (*)())((id (*)(id, SEL))(void *)objc_msgSend)((id)((void (*)())&__main_block_impl_2((void *)__main_block_func_2, &__main_block_desc_2_DATA, age)), sel_registerName("copy"));
        
        NSLog((NSString *)&__NSConstantStringImpl__var_folders_c4_8r5g_7xx6szc_1fms0527v480000gn_T_main_46e938_mi_5,((Class (*)(id, SEL))(void *)objc_msgSend)((id)block1, sel_registerName("class")),((Class (*)(id, SEL))(void *)objc_msgSend)((id)block2, sel_registerName("class")),((Class (*)(id, SEL))(void *)objc_msgSend)((id)block3, sel_registerName("class")));
    }
    return 0;
}
static struct IMAGE_INFO { unsigned version; unsigned flag; } _OBJC_IMAGE_INFO = { 0, 2 };

File: 1、Block原理/Block堆栈对变量捕获的影响/main.m
//
//  main.m
//  Block堆栈对变量捕获的影响
//
//  Created by 林域 on 2021/8/23.
//

#import <Foundation/Foundation.h>
#import "TestObj.h"

int gloable = 5;

int main(int argc, const char * argv[]) {
    
    @autoreleasepool {
                        
        /**
         MAR模式下：
         1、访问了auto变量的block不会被自动拷贝到堆上，需要手动copy，此时block会被copy到堆上，包括引用到的auto变量
         */
        
        //开启 MRC
        void (^block1)(void) = ^(){ // 没有访问任何变量，是一个全局的block
            NSLog(@"block1");
        };
        
        int age = 10;
        void (^block2)(void) = ^(){ //MRC模式下编译器不会自动为访问auto变量的block做copy操作，所以这里还是在栈区
            NSLog(@"block2 = age -- %d -- %p",age,&age); //age是值传递，传进去的是副本，但也仍在栈上。
        };
        
        //之所以要将访问auto变量的block copy到堆区，是因为auto变量存在栈上。由编译器释放。所以如果block在其他地方被延时调用了。此时auto变量可能已经被释放掉了。
        
        NSLog(@"age ad -- %p - %p",&age,&block1);//可以看出 age和block2都在栈上
        //⚠️：变量也会随着block的copy被拷贝到堆区上
        void (^block3)(void) = [ ^(){
            NSLog(@"block3 age -- %d -- %p",age,&age);//从内存地址可以看出age被copy到了堆上
        } copy];
        
        block2();
        block3();
        NSLog(@"block1:%@---->block2:%@----->block3:%@",[block1 class],[block2 class],[block3 class]);
    }
    return 0;
}


